### ISSUE TITLE
Keyboard accessibility

### ISSUE THREAD

**Issue by nickcolley** [FormidableLabs/runpkg#65]
Heya, firstly this is a really useful tool thanks I can see myself using it a lot.

If you wanted to make it more inclusive and usable you could consider doing first improving the keyboard accessibility, which would allow users who cannot use a mouse to access the navigation.

First thing you can try is to tab through the application, can you access the things that you'd click with your mouse?

At the moment you can only tab to the information links.

This is because the anchors do not have a `href` attribute, so they are not focussable elements.

I've also noticed that there is the `order` property used with flexbox for the navigation, if you  make the links focussable you'll notice that tabbing through the navigation will result in you shooting to the bottom of the page instead of going top to bottom as you might expect.

There may be a way to sort the navigation using JavaScript before rendering them.

There's likely more than this to make it accessible to people but solving these issues could be a good first step.

You can also consider:

1. Running an automated testing tool like axe: https://github.com/dequelabs/axe-core
2. Running a screen reader and trying to access the content you'd normally access, with keyboard only: https://accessibility.blog.gov.uk/2018/09/27/assistive-technology-tools-you-can-use-at-no-cost/

Lemme know if you have any questions, thanks again for the tool. 💌

**Comment by VirtualDOMinic** [comment]
Thank you so much for all this feedback @nickcolley - very useful! Great to get feedback from someone involved with GDS, too, as I know the awesome reputation gov.uk has for accessibility.

We should have done a proper Accessibility audit for the MVP itself, but I'll at least be making a number of accessibility improvements priority to make up for this.

Do you have any thoughts on how to treat the code display section? We have an `aria-hidden` `pre` with multiple `spans` that display the syntax-highlighted code, and then a `textarea` that can be tabbed to, but I can't seem to tab out of it to continue to the right-hand static analysis panel. If I give it a `tabindex` of `-1` I can avoid this issue, but this - of course - stops a user from being able to tab to it... This is on Chrome on mac with a branch that has `hrefs` for the anchors in the list, making them tabbable.

**Comment by alexlande** [comment]
@VirtualDOMinic since the `textarea` isn't interactive, it's probably best to use`tabindex="-1"` so that it isn't focusable. It seems like it should be a `pre` ideally, but I see that it's a `textarea` because of the underlying library. If that can't be swapped out yet, you might be able to make do by adding a `role` attribute so that screen readers don't interpret it as a textbox. Not sure if there are any great solutions, but `role='article'` (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Article_Role) might be sort of close.

In general, you can avoid the focus trap when you tab into `react-simple-code-editor` with the `ignoreTabKey` prop. I use that pretty much any time I use `react-live` for the same reason.

**Comment by nickcolley** [comment]
@VirtualDOMinic like Alex has said, first focus on using semantic HTML. Users of assistive technologies have other ways to access prose content, or in this case code.

Remember that if you're new to a screen reader, there can be a good chance you don't know how to operate it, rather than there being a problem with your interface.

So focus on making what you do compatible with screen readers, and any usability considerations be mindful of.

If you really want to know if your tool works for disabled users, you could include them in your user research.

Given that I imagine you don't have budget for that on this kind of project perhaps once you've done everything you think to remove barriers you could tweet it out and see if you can get any feedback from people...

**Comment by kiraarghy** [comment]
Hey all, whilst @VirtualDOMinic is implementing accessibility in his PR I've added Axe testing to our CI

https://github.com/FormidableLabs/runpkg/pull/75

**Comment by VirtualDOMinic** [comment]
PR #69 has been updated with more (what I believe to be) improvements! Taking off the "WIP" tag but will leave the PR un-merged for a little under a day in case there's more feedback.

This won't be the final PR for improving accessibility, though. Just the start.

**Comment by lukejacksonn** [comment]
In the V1 release we worked hard to make sure every interaction was tab-able (I know is just scratching the surface of a11y) which is a good start. We might have regressed in other areas though. Happy to fix more specific issues if anyone raises them.

Closing here in the meantime.

**Comment by nickcolley** [comment]
@lukejacksonn thanks for taking this seriously and making improvements. ❤️
